### PR TITLE
Python: Customizable server application

### PR DESCRIPTION
**Pull request by mrjerryjohns**
This adds support for launching the REPL in 'server' mode that then creates a Matter compliant, commissionable server application that is highly customizable. An endpoint 0 is provided that contains the usual bevy of utility clusters. A `chip.dynamic_server.Server` class is provided that can create any number of custom application endpoints composed of clusters selected from the Python cluster object archive.

A state store is exposed using cluster objects that can be mutated using APIs on the `Server` class that will reflect correctly over the IM.

Future updates here will include support for auto-cluster instantiation using code-generated device types, custom command handlers that can be registered for a given command on a given cluster instance.

APIs are fairly straightforward:
* `Server.CreateEndpoint(..)` creates an endpoint given a list of clusters, and automatically sets up a state store with the right attributes organized in a hierarchical fashion. You can then alter the state of attributes, or mark an optional attribute as valid and implemented.
* `Server.FinalizeEndpoint(..)` actually 'commits' the implemented attributes and makes the endpoint live. If you read the endpoint remotely over the IM, it will now appear.
* `Server.SetAttribute(..)` permits writing to the attribute value in the store. This is reflected immediately over the IM.

More details:
* Significant fixes to `ServerInit.cpp` to now mirror the fix-ups made on the controller side earlier, as well as re-use of the common stack initialization logic.
* Consolidation of both controller and server initialization modalities in `ChipStack`.

### Testing

Launch a server instance, commission it using another REPL instance, and validate that endpoints can be created and reflected correctly on the controller side using IM reads and subscriptions.

### CODE REVIEW

**Comment by mrjerryjohns** [comment]
Filed #20660 to track the addition of tests to validate these new APIs. It's a bit of work, so will add it as a follow-up.

File: src/controller/python/data-model/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import("${chip_root}/src/app/chip_data_model.gni")
+
+chip_data_model("data-model") {
+  zap_file = "repl.zap"
+
+  zap_pregenerated_dir = "${chip_root}/zzz_generated/repl/zap-generated"
+
+  is_server = true
+
+  #allow_circular_includes_from = [ "${chip_root}/src/controller" ]
```
Comment (tehampson) [inline src/controller/python/data-model/BUILD.gn:29]: Nit should you removed this line?

File: src/controller/python/chip/dynamic_server/Server.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+using GetAttributeValueCb = void (*)(void * appContext, EndpointId endpointId, ClusterId clusterId, AttributeId attributeId,
+                                     char * value, uint16_t * size);
+
+GetAttributeValueCb gGetValueCb;
+void * gGetAttributeContext = nullptr;
+
+} // extern "C"
+
+Controller::Python::CommissionableDataProvider gCommissionableDataProvider;
+chip::app::DefaultAclStorage gAclStorage;
```
Comment (tehampson) [inline src/controller/python/chip/dynamic_server/Server.cpp:85]: I think `gAclStorage` and `sStorageAdapter` can be removed

**Review by woody-apple — APPROVED** [review]

File: scripts/build_python.sh [hunk 3]
```diff
@@ -59,6 +60,7 @@ Input Options:
 
   -t --time_between_case_retries MRPActiveRetryInterval     Specify MRPActiveRetryInterval value
                                                             Default is 300 ms
+  -a --extra_args extra_args                                Extra GN args
```
Comment (Arkq) [inline scripts/build_python.sh:63]:
```suggestion
  -t, --time_between_case_retries MRPActiveRetryInterval    Specify MRPActiveRetryInterval value
                                                            Default is 300 ms
  -a, --extra_args extra_args                               Extra GN args
```

**Comment by turon** [comment]
/rebase
Also, please update commit message from "WIP".

**Pull request closed without merging**